domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt

```diff
@@ -0,0 +1,182 @@
+          }
+          Spotlight.FeatureCase.PROMOTED_STORIES -> {
+            it.promotedStories
+          }
+          Spotlight.FeatureCase.LESSONS_BACK_BUTTON -> {
+            it.lessonsBackButton
+          }
+          Spotlight.FeatureCase.VOICEOVER_PLAY_ICON -> {
+            it.voiceoverPlayIcon
+          }
+          Spotlight.FeatureCase.VOICEOVER_LANGUAGE_ICON -> {
+            it.voiceoverLanguageIcon
+          }
+          Spotlight.FeatureCase.FEATURE_NOT_SET -> {
+            return@transformAsync AsyncResult.Failure(
+              SpotlightFeatureNotFoundException("Spotlight feature was not found")
+            )
+          }
+        }
+        AsyncResult.Success(viewState)
+      }
+  }
+
+  private fun recordSpotlightStateAsync(
+    profileId: ProfileId,
+    feature: Spotlight.FeatureCase,
+    viewState: SpotlightViewState
+  ): Deferred<Any> {
+    return retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(
+      updateInMemoryCache = true
+    ) {
+      val spotlightStateDatabaseBuilder = it.toBuilder()
+
+      val spotlight = when (feature) {
+        Spotlight.FeatureCase.ONBOARDING_NEXT_BUTTON -> {
+          spotlightStateDatabaseBuilder.setOnboardingNextButton(viewState)
+        }
+        Spotlight.FeatureCase.TOPIC_LESSON_TAB -> {
+          spotlightStateDatabaseBuilder.setTopicLessonTab(viewState)
+        }
+        Spotlight.FeatureCase.TOPIC_REVISION_TAB -> {
+          spotlightStateDatabaseBuilder.setTopicRevisionTab(viewState)
+        }
+        Spotlight.FeatureCase.FIRST_CHAPTER -> {
+          spotlightStateDatabaseBuilder.setFirstChapter(viewState)
+        }
+        Spotlight.FeatureCase.PROMOTED_STORIES -> {
+          spotlightStateDatabaseBuilder.setPromotedStories(viewState)
+        }
+        Spotlight.FeatureCase.LESSONS_BACK_BUTTON -> {
+          spotlightStateDatabaseBuilder.setLessonsBackButton(viewState)
+        }
+        Spotlight.FeatureCase.VOICEOVER_PLAY_ICON -> {
+          spotlightStateDatabaseBuilder.setVoiceoverPlayIcon(viewState)
+        }
+        Spotlight.FeatureCase.VOICEOVER_LANGUAGE_ICON -> {
+          spotlightStateDatabaseBuilder.setVoiceoverLanguageIcon(viewState)
+        }
+        Spotlight.FeatureCase.FEATURE_NOT_SET -> {
+          throw SpotlightFeatureNotFoundException("Spotlight feature was not found")
+        }
+      }
+      val spotlightStateDatabase = spotlightStateDatabaseBuilder.build()
+      Pair(spotlightStateDatabase, spotlight)
+    }
+  }
+
+  private fun retrieveCacheStore(
+    profileId: ProfileId
+  ): PersistentCacheStore<SpotlightStateDatabase> {
+    val cacheStore = if (profileId in cacheStoreMap) {
+      cacheStoreMap[profileId]!!
+    } else {
+      val cacheStore =
+        cacheStoreFactory.createPerProfile(
+          CACHE_NAME,
+          SpotlightStateDatabase.getDefaultInstance(),
+          profileId
+        )
+      cacheStoreMap[profileId] = cacheStore
+      cacheStore
+    }
```

Comment:
I suspect ``getOrPut`` would be a more idiomatic way to do this. Consider:

```suggestion
    val cacheStore = cacheStoreMap.getOrPut(profileId) {
        cacheStoreFactory.createPerProfile(
          CACHE_NAME,
          SpotlightStateDatabase.getDefaultInstance(),
          profileId
        )
    }
```

Reply:
done